protocols/kad/src/topology.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use kbucket::KBucketsPeerId;
+use libp2p_core::{Multiaddr, PeerId, topology::MemoryTopology, topology::Topology};
+use multihash::Multihash;
+use protocol::KadConnectionType;
+use std::vec;
+
+/// Trait allowing retreival of information necessary for the Kadmelia system to work.
+pub trait KademliaTopology: Topology {
```

Comment:
Why the `Topology` constraint? `KademliaTopology` does not depend on anything in `Topology`.

Reply:
The Kademlia behaviour needs to be able to call `addresses_of_peer` on Topology in order to answer Kademlia requests from remotes.

Reply:
Yes, by requiring both bounds: `where TTopology: Topology + KademliaTopology`

Reply:
`KademliaTopology` is a trait specifically for topologies that can be used by the Kademlia behaviour, hence why it depends on `Topology`.

Reply:
Otherwise, by following that logic, we should have four traits, `AddKadDiscoveredAddress`, `ClosestPeer`, `AddProvider` and `GetProviders`, and we would require `AddKadDiscoveredAddress + ClosestPeer + AddProvider + GetProviders + Topology`.